react/components/EditorContainer/Sidebar/BlockEditor/hooks.ts

```diff
@@ -0,0 +1,318 @@
+import { equals, path } from 'ramda'
+import { useCallback } from 'react'
+import { defineMessages } from 'react-intl'
+
+import {
+  getActiveContentId,
+  getSchemaPropsOrContent,
+  updateExtensionFromForm,
+} from '../../../../utils/components'
+import { useEditorContext } from '../../../EditorContext'
+import { NEW_CONFIGURATION_ID } from '../consts'
+import { useFormMetaContext } from '../FormMetaContext'
+import { useModalContext } from '../ModalContext'
+import { getFormData } from '../utils'
+
+import { UseFormHandlers } from './typings'
+import {
+  getDefaultConfiguration,
+  omitUndefined,
+  throttledUpdateExtensionFromForm,
+} from './utils'
+
+const messages = defineMessages({
+  saveError: {
+    defaultMessage: 'Something went wrong. Please try again.',
+    id: 'admin/pages.editor.components.content.save.error',
+  },
+  saveSuccess: {
+    defaultMessage: 'Content saved.',
+    id: 'admin/pages.editor.components.content.save.success',
+  },
+})
+
+export const useFormHandlers: UseFormHandlers = ({
+  iframeRuntime,
+  intl,
+  saveContent,
+  setState,
+  showToast,
+  state,
+}) => {
+  const editor = useEditorContext()
+  const formMeta = useFormMetaContext()
+  const modal = useModalContext()
+
+  const handleConditionChange = useCallback(
+    (changes: Partial<typeof state['condition']>) => {
+      setState({
+        condition: {
+          ...(state.condition as ExtensionConfiguration['condition']),
+          ...changes,
+        },
+      })
+
+      if (!formMeta.getWasModified()) {
+        formMeta.setWasModified(true)
+      }
+    },
+    [formMeta, setState, state]
+  )
+
+  const handleFormChange = useCallback(
+    event => {
+      if (
+        state.formData &&
+        !formMeta.getWasModified() &&
+        !equals(omitUndefined(state.formData), omitUndefined(event.formData))
+      ) {
+        formMeta.setWasModified(true)
+      }
+
+      if (event.formData) {
+        setState({ formData: event.formData })
+
+        throttledUpdateExtensionFromForm({
+          data: event.formData,
+          isContent: true,
+          runtime: iframeRuntime,
+          treePath: editor.editTreePath,
+        })
+      }
+    },
+    [editor.editTreePath, formMeta, iframeRuntime, setState, state]
+  )
+
+  const handleFormClose = useCallback(async () => {
+    if (!editor.getIsLoading()) {
+      editor.setIsLoading(true)
+    }
+
+    await iframeRuntime.updateRuntime()
```

Comment:
We're always updating the Runtime for consistency. I've added the loader because it may take a few hundred milliseconds—or, sometimes, even a few seconds—, which could be interpreted as unresponsiveness.